app/code/Magento/InventorySales/Plugin/CatalogInventory/StockManagement/ProcessRegisterProductsSalePlugin.php

```diff
@@ -85,19 +103,26 @@ public function aroundRegisterProductsSale(StockManagement $subject, callable $p
             //TODO: Do we need to throw exception?
             throw new LocalizedException(__('$websiteId parameter is required'));
         }
+
         $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
         $productSkus = $this->getSkusByProductIds->execute(array_keys($items));
-
         $this->checkItemsQuantity($items, $productSkus, $stockId);
+
+        $productTypes = $this->getProductTypesBySkus->execute(array_values($productSkus));
         $reservations = [];
         foreach ($productSkus as $productId => $sku) {
+            if (!$this->isSourceItemsAllowedForProductType->execute($productTypes[$sku])) {
```

Comment:
Why do we need this check?
Do you discover situation when passed sku is related to complex products?

Reply:
This issue already exists
@nmalevanec working on it by issue